### PR TITLE
librbd: error messages encountered disabling object-map while flatten is in-progress

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

File: src/librbd/operation/SnapshotUnprotectRequest.cc [hunk 1]
```diff
@@ -259,7 +259,7 @@ void SnapshotUnprotectRequest<I>::send_scan_pool_children() {
     boost::lambda::bind(boost::lambda::new_ptr<C_ScanPoolChildren<I> >(),
       boost::lambda::_1, &image_ctx, pspec, pools, boost::lambda::_2));
   AsyncObjectThrottle<I> *throttle = new AsyncObjectThrottle<I>(
-    this, image_ctx, context_factory, ctx, NULL, 0, pools.size());
+    nullptr, image_ctx, context_factory, ctx, NULL, 0, pools.size());
```
Comment (jdurgin) [inline src/librbd/operation/SnapshotUnprotectRequest.cc]: this line is a little too implicit to my conference-addled brain - why include it?

Comment (dillaman) [inline src/librbd/operation/SnapshotUnprotectRequest.cc]: black magic -- by not passing a pointer to the AsyncRequest object (SnapshotUnprotectRequest in this case), the iteration over the pools cannot be canceled in-flight.  Unlike a trim flatten, it should be a relatively quick operation.

Comment (jdurgin) [inline src/librbd/operation/SnapshotUnprotectRequest.cc]: makes sense, I was misreading git grep. Flatten does do this, and copyup indeed does not need to. lgtm

**Comment by trociny** [comment]
@dillaman I am observing this recursive lock when running rbd_mirror.sh:
```
recursive lock of librbd::ImageCtx::owner_lock (0x7fea34002010) (17)
 ceph version 10.2.0-338-ga543eb4 (a543eb4b0475ac29a3e15c3218188a6c243e8021)
 1: (()+0x25d908) [0x7feacfa44908]
 2: (()+0x25e2e5) [0x7feacfa452e5]
 3: (Context::complete(int)+0x9) [0x7feacf999e39]
 4: (librbd::ImageState<librbd::ImageCtx>::complete_action_unlock(librbd::ImageState<librbd::ImageCtx>::State, int)+0x10c) [0x7f
 5: (librbd::ImageState<librbd::ImageCtx>::handle_refresh(int)+0xac) [0x7feacf9ed4bc]
 6: (Context::complete(int)+0x9) [0x7feacf999e39]
 7: (ThreadPool::worker(ThreadPool::WorkThread*)+0xa9f) [0x7feacfb89b8f]
 8: (ThreadPool::WorkThread::entry()+0x10) [0x7feacfb8aab0]
 9: (()+0x80a4) [0x7feac55ec0a4]
 10: (clone()+0x6d) [0x7feac3f5404d]
```
gdb:
```
#7  0x00007feacfc02d55 in lockdep_will_lock (name=0x7fea34002ac8 "librbd::ImageCtx::owner_lock (0x7fea34002010)", 
    id=<optimized out>, force_backtrace=force_backtrace@entry=false) at common/lockdep.cc:259
#8  0x00007feacf9cfc56 in RWLock::get_read (this=0x7fea340021a8) at ./common/RWLock.h:99
#9  0x00007feacfa44908 in RLocker (lock=..., this=<synthetic pointer>) at ./common/RWLock.h:161
#10 librbd::(anonymous namespace)::C_InvokeAsyncRequest<librbd::ImageCtx>::send_acquire_exclusive_lock (
    this=this@entry=0x7fea3400ffe0) at librbd/Operations.cc:155
#11 0x00007feacfa452e5 in librbd::(anonymous namespace)::C_InvokeAsyncRequest<librbd::ImageCtx>::handle_refresh_image (
    this=0x7fea3400ffe0, r=0) at librbd/Operations.cc:151
#12 0x00007feacf999e39 in Context::complete (this=0x7fea34004f20, r=<optimized out>) at ./include/Context.h:64
#13 0x00007feacf9ecb3c in librbd::ImageState<librbd::ImageCtx>::complete_action_unlock (this=this@entry=0x7fea34005f60, 
    next_state=next_state@entry=librbd::ImageState<librbd::ImageCtx>::STATE_OPEN, r=r@entry=0) at librbd/ImageState.cc:247
#14 0x00007feacf9ed4bc in librbd::ImageState<librbd::ImageCtx>::handle_refresh (this=0x7fea34005f60, r=0)
    at librbd/ImageState.cc:353
#15 0x00007feacf999e39 in Context::complete (this=0x7fea34019f80, r=<optimized out>) at ./include/Context.h:64
#16 0x00007feacfb89b8f in ThreadPool::worker (this=0x7fea340066e0, wt=0x7fea34006b80) at common/WorkQueue.cc:128
#17 0x00007feacfb8aab0 in ThreadPool::WorkThread::entry (this=<optimized out>) at common/WorkQueue.h:445
#18 0x00007feac55ec0a4 in start_thread (arg=0x7fea6affd700) at pthread_create.c:309
#19 0x00007feac3f5404d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```
It looks like due to 9cf6821

**Pull request merged**